pkg/cmd/image/list.go

```diff
@@ -0,0 +1,407 @@
+			if err != nil {
+				return err
+			}
+		}
+
+		imageList = imgutil.FilterImages(imageList, beforeImages, sinceImages)
+	}
+	return printImages(ctx, options, client, imageList)
+}
+
+func filterByReference(imageList []images.Image, filters []string) ([]images.Image, error) {
+	var filteredImageList []images.Image
+	logrus.Debug(filters)
+	for _, image := range imageList {
+		logrus.Debug(image.Name)
+		var matches int
+		for _, f := range filters {
+			var ref dockerreference.Reference
+			var err error
+			ref, err = dockerreference.ParseAnyReference(image.Name)
+			if err != nil {
+				return nil, fmt.Errorf("unable to parse image name: %s while filtering by reference because of %s", image.Name, err.Error())
+			}
+
+			familiarMatch, err := dockerreference.FamiliarMatch(f, ref)
+			if err != nil {
+				return nil, err
+			}
+			regexpMatch, err := regexp.MatchString(f, image.Name)
+			if err != nil {
+				return nil, err
+			}
+			if familiarMatch || regexpMatch {
+				matches++
+			}
+		}
+		if matches == len(filters) {
+			filteredImageList = append(filteredImageList, image)
+		}
+	}
+	return filteredImageList, nil
+}
+
+func filterDangling(imageList []images.Image, dangling bool) []images.Image {
+	var filtered []images.Image
+	for _, image := range imageList {
+		_, tag := imgutil.ParseRepoTag(image.Name)
+
+		if dangling && tag == "" {
+			filtered = append(filtered, image)
+		}
+		if !dangling && tag != "" {
+			filtered = append(filtered, image)
+		}
+	}
+	return filtered
+}
+
+func filterByLabel(ctx context.Context, client *containerd.Client, imageList []images.Image, filters map[string]string) ([]images.Image, error) {
+	for lk, lv := range filters {
+		var imageLabels []images.Image
+		for _, img := range imageList {
+			ci := containerd.NewImage(client, img)
+			cfg, _, err := imgutil.ReadImageConfig(ctx, ci)
+			if err != nil {
+				return nil, err
+			}
+			if val, ok := cfg.Config.Labels[lk]; ok {
+				if val == lv || lv == "" {
+					imageLabels = append(imageLabels, img)
+				}
+			}
+		}
+		imageList = imageLabels
+	}
+	return imageList, nil
+}
+
+type imagePrintable struct {
+	// TODO: "Containers"
+	CreatedAt    string
+	CreatedSince string
+	Digest       string // "<none>" or image target digest (i.e., index digest or manifest digest)
+	ID           string // image target digest (not config digest, unlike Docker), or its short form
+	Repository   string
+	Tag          string // "<none>" or tag
+	Name         string // image name
+	Size         string // the size of the unpacked snapshots.
+	BlobSize     string // the size of the blobs in the content store (nerdctl extension)
+	// TODO: "SharedSize", "UniqueSize", "VirtualSize"
+	Platform string // nerdctl extension
+}
+
+func printImages(ctx context.Context, options types.ImagesCommandOptions, client *containerd.Client, imageList []images.Image) error {
+	digestsFlag := options.Digests
+	var w io.Writer
+	w = os.Stdout
```

Comment:
I know this is original code, but later we might pass a writer to this func instead of binding to os.Stdout

Reply:
SGTM, maybe another PR could be better